persist/fs/commitlog/commit_log.go

```diff
@@ -0,0 +1,275 @@
+			if err := l.openWriter(now); err != nil {
+				l.metrics.IncCounter("writes.errors", 1)
+				l.metrics.IncCounter("writes.openerrors", 1)
+				// Re-attempt next write
+				retry = write
+				continue
+			}
+		}
+
+		// For writes requiring acks add to pending acks
+		if write.completionFn != nil {
+			l.pendingAcks = append(l.pendingAcks, write.completionFn)
+		}
+
+		err := l.writer.Write(write.series, write.datapoint, write.unit, write.annotation)
+		if err != nil {
+			l.metrics.IncCounter("writes.errors", 1)
+			// Log error and expire writer
+			l.writeError(err)
+			// Re-attempt next write
+			retry = write
+			continue
+		}
+
+		l.metrics.IncCounter("writes", 1)
+	}
+
+	if l.writer != nil {
+		writer := l.writer
+		l.writer = nil
+		l.closeErr <- writer.Close()
+	} else {
+		l.closeErr <- nil
+	}
+}
+
+func (l *commitLog) onFlush(err error) {
+	// Taking the pending acks is safe here as is called by the goroutine
+	// enqueuing the pending acks in write method from the writer as a
+	// result of a write
+	if len(l.pendingAcks) == 0 {
+		return
+	}
+
+	// Safe to callback acks as all they do is record a result and unblock
+	// another goroutine
+	for i := range l.pendingAcks {
+		l.pendingAcks[i](err)
+	}
+	l.pendingAcks = l.pendingAcks[:0]
+}
+
+func (l *commitLog) openWriter(now time.Time) error {
+	log := l.opts.GetLogger()
+	if l.writer != nil {
+		if err := l.writer.Close(); err != nil {
+			log.Errorf("failed to close commit log: %v", err)
+			// If we failed to close then create a new commit log writer
+			l.writer = nil
+		}
+	}
+	if l.writer == nil {
+		l.writer = newCommitLogWriter(l.onFlush, l.opts)
+	}
+
+	blockSize := l.opts.GetBlockSize()
+	start := now.Truncate(blockSize)
+	if err := l.writer.Open(start, blockSize); err != nil {
+		log.Errorf("failed to open new commit log: %v", err)
+		return err
+	}
+
+	l.writerExpireAt = start.Add(blockSize)
+	return nil
+}
+
+func (l *commitLog) writeError(err error) {
+	log := l.opts.GetLogger()
+	log.Errorf("failed to write commit log entry: %v", err)
+	// Explicitly expire the writer so we reopen a new commit log
+	l.writerExpireAt = timeZero
+}
+
+func (l *commitLog) Write(
+	series m3db.CommitLogSeries,
+	datapoint m3db.Datapoint,
+	unit xtime.Unit,
+	annotation []byte,
+) error {
+	l.RLock()
+	if l.closed {
+		l.RUnlock()
+		return errCommitLogClosed
+	}
+
+	var (
+		wg     sync.WaitGroup
+		result error
+	)
+	wg.Add(1)
+	completion := func(err error) {
+		result = err
+		wg.Done()
+	}
+	l.writes <- commitLogWrite{
+		series:       series,
+		datapoint:    datapoint,
+		unit:         unit,
+		annotation:   annotation,
+		completionFn: completion,
+	}
+	l.RUnlock()
+
+	wg.Wait()
+
+	return result
+}
+
+func (l *commitLog) WriteBehind(
+	series m3db.CommitLogSeries,
+	datapoint m3db.Datapoint,
+	unit xtime.Unit,
+	annotation []byte,
+) error {
+	l.RLock()
+	if l.closed {
+		l.RUnlock()
+		return errCommitLogClosed
+	}
+	l.writes <- commitLogWrite{
```

Comment:
WriteBehind will still block when the buffered chan is full right? Might be worth putting this in a select and return some sort of error or kick this off in a go-routine so that this call is never blocking.